src/tests/units/html_tests.cpp

```diff
@@ -0,0 +1,493 @@
+  std::string html_code("<p>This text &amp; has <b>HTML</b> in it</p>");
+
+  std::string input(html_code);
+  HTML html(std::move(input), false);
+  CHECK(input == html_code);
+
+  Response response;
+  response.source.text = html_code;
+  response.target.text = html_code;
+  html.Restore(response);
+
+  // Assert that Restore() does not mess with my HTML code
+  CHECK(response.source.text == html_code);
+}
+
+TEST_CASE("Test reconstruction") {
+  std::string input("<p><input>H<u>e</u>llo <b>world</b> how <u>are you</u>?</p>\n");
+
+  std::string text(input);
+  HTML html(std::move(text), true);  // TODO: move, but really a reference?
+  CHECK(text == "Hello world how are you?\n");
+
+  AnnotatedText source(std::move(text));
+  std::vector<string_view> tokens{
+      string_view(source.text.data() + 0, 4),   // Hell
+      string_view(source.text.data() + 4, 1),   // o
+      string_view(source.text.data() + 5, 6),   // _world
+      string_view(source.text.data() + 11, 4),  // _how
+      string_view(source.text.data() + 15, 4),  // _are
+      string_view(source.text.data() + 19, 4),  // _you
+      string_view(source.text.data() + 23, 1),  // ?
+      string_view(source.text.data() + 24, 0),  // "\n" (but 0 length?)
+  };
+
+  source.recordExistingSentence(tokens.begin(), tokens.end(), source.text.data());
+
+  Response response;
+  response.source = source;
+
+  html.Restore(response);
+  // CHECK(response.source.text == input);  // fails because <u></u> has been moved to the front of the token
+  CHECK(response.source.text == "<p><input><u></u>Hello <b>world</b> how <u>are you</u>?</p>\n");
+
+  std::vector<ByteRange> restored_tokens{
+      ByteRange{0, 0 + 0},    // (start of sentence)
+      ByteRange{0, 0 + 21},   // <p><input>H<u>e</u>ll
+      ByteRange{21, 21 + 1},  // o
+      ByteRange{22, 22 + 9},  // _<b>world
+      ByteRange{31, 31 + 8},  // </b>_how
+      ByteRange{39, 39 + 7},  // _<u>are
+      ByteRange{46, 46 + 4},  // _you
+      ByteRange{50, 50 + 5},  // </u>?
+      ByteRange{55, 55 + 0},  // ""
+      ByteRange{55, 55 + 5},  // </p>\n
+  };
+  CHECK(response.source.text.size() == restored_tokens.back().end);
+  CHECK(AsByteRanges(response.source) == restored_tokens);
+
+  // Same test as above, but easier to read. Will use this further down.
+  std::vector<std::string> restored_tokens_str{"",
+                                               "<p><input><u></u>Hell",  // Should really be "<p><input>H<u>e</u>ll"
+                                               "o",
+                                               " <b>world",
+                                               "</b> how",
+                                               " <u>are",
+                                               " you",
+                                               "</u>?",
+                                               "",  // end of sentence
+                                               "</p>\n"};
```

Comment:
Okay my bad, I already have some logic around making empty elements disappear. But in this case you can see that the `<u></u>` became empty because it was just wrapped around the `e`, which makes up only a small part of the `Hell` token. What then happens is that during reconstruction it goes through all tokens, sees which spans match with a token, and then apply the tags from that span to the token. So the token is the smallest level it can edit at. In this case, it would apply `<p>` for the span `H`, then `<p>,<u>` for span `e`, and then revert to just `<p>` for `llo_` as it scans through all spans that cover the token `Hell`.